src/gt4py/next/iterator/transforms/inline_lifts.py

```diff
@@ -138,125 +194,213 @@ class Flag(enum.IntEnum):
+                    reserved_symbol_names + list(new_arg_exprs.keys())
+                ):
+                    new_param = im.sym(arg.id)
+                    trace_shifts.copy_recorded_shifts(
+                        from_=param,
+                        to=new_param,
+                        required=self.inline_single_pos_deref_lift_args_only,
+                    )
+                    new_arg_exprs[new_param] = arg
+
+                    new_arg = im.ref(arg.id)
+                    trace_shifts.copy_recorded_shifts(
+                        from_=param,
+                        to=new_arg,
+                        required=self.inline_single_pos_deref_lift_args_only,
+                    )
+                    assert param not in inlined_args
+                    inlined_args[param] = new_arg
+                else:
+                    new_arg_exprs[param] = arg
+
+        new_stencil_body = im.let(
+            *((param, inlined_lift) for param, inlined_lift in inlined_args.items())  # type: ignore[arg-type] # mypy not smart enough
+        )(stencil.expr)
+
+        # it is likely that some of the let args can be immediately inlined. do this eagerly
+        # here
+        new_stencil_body = inline_lambda(new_stencil_body, opcount_preserving=True)
+
+        assert not self.inline_single_pos_deref_lift_args_only or all(
+            hasattr(param.annex, "recorded_shifts") for param in new_arg_exprs.keys()
+        )
+
+        new_stencil = im.lambda_(*new_arg_exprs.keys())(new_stencil_body)
+        new_applied_lift = im.lift(new_stencil)(*new_arg_exprs.values())
+        trace_shifts.copy_recorded_shifts(from_=node, to=new_applied_lift, required=False)
+        return self.visit(new_applied_lift, **kwargs)
+
+    def visit_FunCall(self, node: ir.FunCall, **kwargs) -> ir.Node:
+        recurse: bool = kwargs["recurse"]
+
+        if recurse:
+            fun_kwargs = {k: v for k, v in kwargs.items() if k != "is_scan_pass_context"}
+            new_node = ir.FunCall(
+                fun=self.generic_visit(node.fun, is_scan_pass_context=_is_scan(node), **fun_kwargs),
+                args=self.generic_visit(node.args, **kwargs),
+            )
+            trace_shifts.copy_recorded_shifts(from_=node, to=new_node, required=False)
+        else:
+            new_node = node
 
-                new_stencil = im.lambda_(*new_arg_exprs.keys())(inlined_call)
-                return im.lift(new_stencil)(*new_arg_exprs.values())
+        for transformation in self.Flag:
+            if self.flags & transformation:
+                assert isinstance(transformation.name, str)
+                method = getattr(self, f"transform_{transformation.name.lower()}")
+                transformed_node = method(new_node, **kwargs)
+                # if the transformation returned `None` it did not apply and we continue.
+                if transformed_node is not None:
+                    return transformed_node
```

Comment:
I am a bit confused by this pattern. The first transformation that matches is applied? Because each transformation matches exactly once?